File: backend-java/src/main/java/com/bffgin/backend/flags/FeatureFlagPoller.java
package com.bffgin.backend.flags;

import javax.sql.DataSource;
import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;

/**
 * feature_flagsテーブルを10秒間隔で直接ポーリングする(bffのようなHTTPポーリングではない、
 * backend-c/backend-cpp/backend-rustと同じ設計)。外部公開API(backend.external-tasks-pagination-v2)
 * の判定に使う。
 *
 * 読み取り側(variation)がロック無しで安全に読めるよう、キャッシュ全体を不変Mapとして
 * volatileフィールドへ丸ごと差し替える(ポーリングスレッドだけが書き込み、他スレッドは
 * 常に一貫したスナップショットを読む。Java実装のVirtual Threadsと同じ「共有可変状態を
 * 極力減らす」設計方針)
 */
public final class FeatureFlagPoller implements AutoCloseable {

    public record Entry(boolean enabled, String defaultVariation) {
    }

    private final DataSource dataSource;
    private final ScheduledExecutorService scheduler = Executors.newSingleThreadScheduledExecutor(
            r -> {
                Thread t = new Thread(r, "feature-flag-poller");
                t.setDaemon(true);
                return t;
            });
    private volatile Map<String, Entry> entries = Map.of();

    public FeatureFlagPoller(DataSource dataSource) {
        this.dataSource = dataSource;
    }

    public void start() {
        pollOnce();
        scheduler.scheduleAtFixedRate(this::pollOnce, 10, 10, TimeUnit.SECONDS);
    }

    public void pollOnce() {
        Map<String, Entry> next = new HashMap<>();
        try (Connection conn = dataSource.getConnection();
                PreparedStatement ps = conn.prepareStatement(
                        "SELECT flag_key, enabled, default_variation FROM feature_flags");
                ResultSet rs = ps.executeQuery()) {
            while (rs.next()) {
                next.put(rs.getString("flag_key"), new Entry(rs.getBoolean("enabled"), rs.getString("default_variation")));
            }
            entries = Map.copyOf(next);
        } catch (SQLException e) {
            // 【バッド/グッドプラクティス】1回のポーリング失敗で例外を伝播させ
            // スケジューラ自体を止めてしまうと、以後永久にフラグが更新されなくなる。
            // ここでは失敗を握りつぶし、既存のキャッシュ(直前の正常な取得結果)を
            // 保持したまま次回のポーリングに委ねる
        }
    }

    /** enabled=falseの場合、または未知のflagKeyの場合はdefaultValueを返す */
    public String variation(String flagKey, String defaultValue) {
        Entry entry = entries.get(flagKey);
        if (entry == null || !entry.enabled()) {
            return defaultValue;
        }
        return entry.defaultVariation();
    }

    @Override
    public void close() {
        scheduler.shutdownNow();
    }
}

File: backend-kotlin/src/main/kotlin/com/bffgin/backend/flags/FeatureFlagPoller.kt
package com.bffgin.backend.flags

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import javax.sql.DataSource

/**
 * feature_flagsテーブルを10秒間隔で直接ポーリングする(bffのようなHTTPポーリングではない、
 * backend-java/backend-rust/backend-cpp/backend-cと同じ設計)。外部公開API
 * (backend.external-tasks-pagination-v2)の判定にのみ使う。
 *
 * 【このKotlin実装の一貫した設計原則をポーラーにも適用する】TaskRepository同様、
 * ここでのJDBC呼び出しも`withContext(Dispatchers.IO)`で明示的に囲む。ポーリングという
 * バックグラウンド処理だからといって、この設計原則(ブロッキングI/Oは呼び出し側が
 * 自己申告してディスパッチャを明示的に切り替える)の例外にはしない。
 *
 * キャッシュは`@Volatile`な不変Mapを毎回まるごと置き換える方式。JVMメモリモデル上、
 * `@Volatile`フィールドへの書き込みはそれ以前の全ての書き込みと同期し、他スレッドから見て
 * 安全に公開される(safe publication)ため、読み取り側にロックは不要
 */
class FeatureFlagPoller(private val dataSource: DataSource) {

    private data class Entry(val enabled: Boolean, val defaultVariation: String)

    @Volatile
    private var cache: Map<String, Entry> = emptyMap()

    private var job: Job? = null

    fun start(scope: CoroutineScope) {
        job = scope.launch {
            while (isActive) {
                pollOnce()
                delay(10_000)
            }
        }
    }

    fun stop() {
        job?.cancel()
    }

    /** enabled=trueかつ見つかった場合はdefault_variation、それ以外は呼び出し側のdefaultValue */
    fun variation(flagKey: String, defaultValue: String): String {
        val entry = cache[flagKey] ?: return defaultValue
        return if (entry.enabled) entry.defaultVariation else defaultValue
    }

    /** 通常は10秒間隔の内部ループから呼ばれるが、テストでは即時実行のために直接呼び出す(公開) */
    suspend fun pollOnce() {
        val newCache = withContext(Dispatchers.IO) {
            val result = mutableMapOf<String, Entry>()
            dataSource.connection.use { conn ->
                conn.prepareStatement("SELECT flag_key, enabled, default_variation FROM feature_flags").use { ps ->
                    ps.executeQuery().use { rs ->
                        while (rs.next()) {
                            result[rs.getString("flag_key")] = Entry(rs.getBoolean("enabled"), rs.getString("default_variation"))
                        }
                    }
                }
            }
            result
        }
        cache = newCache
    }
}
